### PR TITLE
Stash v1beta1 E2E test for PVC 

### PR DESCRIPTION
**Pull request by suaas21**
xref: #707

Tasks:

- [x] Volume backup and restore for PVC

### CODE REVIEW

File: test/e2e/volume_test.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+			Skip("Missing repository credential")
+		}
+		pvc = f.GetPersistentVolumeClaim()
+		err = f.CreatePersistentVolumeClaim(pvc)
+		Expect(err).NotTo(HaveOccurred())
+		repo = f.Repository(cred.Name, pvc.Name)
+		fmt.Println("pvc for repo", pvc.Name)
+		targetref = v1beta1.TargetRef{}
+		backupCfg = f.BackupConfiguration(repo.Name, targetref)
+		backupCfg.Spec.Target = &v1beta1.Target{
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: Move this part to a function and then just use:
```
backupCfg.Spec.Target = framework.PvcBackupTarget(pvc)
```

You can directly use the function as:
```
backupCfg = f.BackupConfiguration(repo.Name, framework.PvcBackupTarget(pvc))
```

File: test/e2e/volume_test.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+				},
+			},
+			Directories: []string{
+				framework.TestSourceDataMountPath,
+			},
+		}
+		backupCfg.Spec.Task.Name = backupTask.Name
+
+		restoreSession = f.RestoreSession(repo.Name, targetref, rules)
+		restoreSession.Spec.Target = &v1beta1.Target{
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: Same as previous comment.

File: test/e2e/volume_test.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+			err = util.WaitUntilPodRunning(f.KubeClient, pod.ObjectMeta)
+			Expect(err).NotTo(HaveOccurred())
+
+		}
+	)
+	Context("General Backup && Restore for PVC Volume", func() {
+		BeforeEach(func() {
+			bpv = f.GetPersistentVolume()
+			bpvc = f.GetPersistentVolumeClaim()
+			fmt.Println(bpvc.Name)
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: Remove this line.

File: test/e2e/volume_test.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+			updateStatusFunc = f.UpdateStatusFunction()
+			backupFunc = f.BackupFunction()
+			restoreFunc = f.RestoreFunction()
+
+			err = f.CreateFunction(updateStatusFunc)
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: Move creation of `Function` and `Task` in `BeforeEach` of the Context/Describe and reuse them in other cases.

File: test/e2e/volume_test.go [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+			err = framework.WaitUntilRepositoryDeleted(f.StashClient, repo)
+			Expect(err).NotTo(HaveOccurred())
+
+			err = f.DeleteRestoreSession(restoreSession.ObjectMeta)
+			Expect(err).NotTo(HaveOccurred())
+			err = framework.WaitUntilRestoreSessionDeleted(f.StashClient, restoreSession.ObjectMeta)
+			Expect(err).NotTo(HaveOccurred())
+
+		})
+		FIt("General Backup new PVC", func() {
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: Remove focus -> `FIt`.

File: test/e2e/volume_test.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+
+			By("Creating Restore Session")
+			restoreSession.Spec.Target.Ref.Name = rpvc.Name
+			err = f.CreateRestoreSession(restoreSession)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Waiting for restore to succeed")
+			f.EventuallyRestoreSessionPhase(restoreSession.ObjectMeta).Should(Equal(v1beta1.RestoreSessionSucceeded))
+
+			By("delete previous Pod")
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: what pod are you deleting?

File: test/e2e/volume_test.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+			err = f.DeletePod(pod.ObjectMeta)
+			Expect(err).NotTo(HaveOccurred())
+			By("Create Pod and Generate sample Data")
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: Why creating pod an generating data again?

File: pkg/util/kubernetes.go [hunk 8]
```diff
@@ -373,3 +374,32 @@ func WaitUntilDeploymentConfigReady(c oc_cs.Interface, meta metav1.ObjectMeta) e
 		return false, nil
 	})
 }
+
+func WaitUntilPodRunning(kubeClient kubernetes.Interface, meta metav1.ObjectMeta) error {
```
Comment (tamalsaha) [inline pkg/util/kubernetes.go]: Why don't we use kutil?

File: test/e2e/volume_test.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+			err = framework.WaitUntilRepositoryDeleted(f.StashClient, repo)
+			Expect(err).NotTo(HaveOccurred())
+
+			err = f.DeleteRestoreSession(restoreSession.ObjectMeta)
+			Expect(err).NotTo(HaveOccurred())
+			err = framework.WaitUntilRestoreSessionDeleted(f.StashClient, restoreSession.ObjectMeta)
+			Expect(err).NotTo(HaveOccurred())
+
+		})
+		FIt("General Backup new PVC", func() {
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: Remove focus from `FIt()`

File: test/e2e/volume_test.go [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+)
+
+var _ = Describe("Volume", func() {
+	BeforeEach(func() {
+		f = root.Invoke()
+
+		updateStatusFunc = f.UpdateStatusFunction()
+		backupFunc = f.PvcBackupFunction()
+		restoreFunc = f.PvcRestoreFunction()
+
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: Add description what happening here. i.e.:
```
By("Creating functions")
```

File: test/e2e/volume_test.go [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+		err = f.CreateFunction(updateStatusFunc)
+		Expect(err).NotTo(HaveOccurred())
+		err = f.CreateFunction(backupFunc)
+		Expect(err).NotTo(HaveOccurred())
+		err = f.CreateFunction(restoreFunc)
+		Expect(err).NotTo(HaveOccurred())
+
+		backupTask = f.BackupTask()
```
Comment (hossainemruz) [inline test/e2e/volume_test.go]: Add `By("Creating Tasks")`

File: test/e2e/framework/task.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+func (f *Invocation) BackupTask() v1beta1.Task {
+	return v1beta1.Task{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: PVCBackupTaskName,
+		},
+		Spec: v1beta1.TaskSpec{
+			Steps: []v1beta1.FunctionRef{
+				{
+					Name: "pvc-backup",
```
Comment (hossainemruz) [inline test/e2e/framework/task.go]: Use constant for`pvc-backup` .
```
 const FunctionPvcBackup = "pvc-backup"
```

File: test/e2e/framework/task.go [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+func (f *Invocation) RestoreTask() v1beta1.Task {
+	return v1beta1.Task{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: PVCRestoreTaskName,
+		},
+		Spec: v1beta1.TaskSpec{
+			Steps: []v1beta1.FunctionRef{
+				{
+					Name: "pvc-restore",
```
Comment (hossainemruz) [inline test/e2e/framework/task.go]: Use constant for`pvc-restore` .
```
 const FunctionPvcRestore = "pvc-restore"
```

File: test/e2e/framework/task.go [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+							Value: tarVolName,
+						},
+						{
+							Name:  secVol,
+							Value: secVolName,
+						},
+					},
+				},
+				{
+					Name: "update-status",
```
Comment (hossainemruz) [inline test/e2e/framework/task.go]: Use constant for`update-status` .
```
 const UpdateStatusFunction = "update-status"
```

File: test/e2e/framework/task.go [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
```
Comment (hossainemruz) [inline test/e2e/framework/task.go]: Use constant for `update-status`.

File: test/e2e/framework/functions.go [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+		Registry: DockerRegistry,
+		Image:    docker.ImageStash,
+		Tag:      DockerImageTag,
+	}.ToContainerImage()
+}
+
+func (f *Invocation) UpdateStatusFunction() v1beta1.Function {
+	return v1beta1.Function{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "update-status",
```
Comment (hossainemruz) [inline test/e2e/framework/functions.go]: Use constant for `update-status`.

File: test/e2e/framework/functions.go [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+				fmt.Sprintf("--enable-status-subresource=${%s:=%s}", apis.StatusSubresourceEnabled, apis.StatusSubresourceEnabled),
+			},
+		},
+	}
+}
+
+func (f *Invocation) PvcBackupFunction() v1beta1.Function {
+	return v1beta1.Function{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "pvc-backup",
```
Comment (hossainemruz) [inline test/e2e/framework/functions.go]: Use constant for `pvc-backup`.

File: test/e2e/framework/functions.go [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+				},
+			},
+		},
+	}
+}
+
+func (f *Invocation) PvcRestoreFunction() v1beta1.Function {
+	return v1beta1.Function{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "pvc-restore",
```
Comment (hossainemruz) [inline test/e2e/framework/functions.go]: Use constant for `pvc-restore`.

**Pull request merged**